packages/scene-composer/src/components/panels/AddComponentMenu.tsx

```diff
@@ -0,0 +1,162 @@
+  Overlay = 'add-component-overlay',
+  DataBinding = 'add-component-data-binding',
+}
+
+type AddComponentMenuItem = ToolbarItemOptions & {
+  uuid: ObjectTypes;
+};
+
+const labelStrings: { [key in ObjectTypes]: MessageDescriptor } = defineMessages({
+  [ObjectTypes.Component]: { defaultMessage: 'Add component', description: 'Menu Item label' },
+  [ObjectTypes.Overlay]: { defaultMessage: 'Overlay', description: 'Menu Item label' },
+  [ObjectTypes.DataBinding]: { defaultMessage: 'Data binding', description: 'Menu Item label' },
+});
+
+const textStrings = defineMessages({
+  [ObjectTypes.Overlay]: { defaultMessage: 'Add overlay', description: 'Menu Item' },
+  [ObjectTypes.DataBinding]: { defaultMessage: 'Add data binding', description: 'Menu Item' },
+});
+
+export const AddComponentMenu: React.FC<AddComponentMenuProps> = ({ onSelect }) => {
+  const sceneComposerId = useContext(sceneComposerIdContext);
+  const addComponentInternal = useStore(sceneComposerId)((state) => state.addComponentInternal);
+  const updateComponentInternal = useStore(sceneComposerId)((state) => state.updateComponentInternal);
+  const selectedSceneNodeRef = useStore(sceneComposerId)((state) => state.selectedSceneNodeRef);
+  const getSceneNodeByRef = useStore(sceneComposerId)((state) => state.getSceneNodeByRef);
+  const { formatMessage } = useIntl();
+  const selectedSceneNode = getSceneNodeByRef(selectedSceneNodeRef);
+  const dataBindingComponentEnabled = getGlobalSettings().featureConfig[COMPOSER_FEATURES.DataBinding];
+
+  const isTagComponent = !!findComponentByType(selectedSceneNode, KnownComponentType.Tag);
+  const isOverlayComponent = !!findComponentByType(selectedSceneNode, KnownComponentType.DataOverlay);
+
+  const mapToMenuItem = useCallback(
+    (item: ToolbarItemOptionRaw): AddComponentMenuItem => {
+      const typeId: ObjectTypes = item.uuid as ObjectTypes;
+      return {
+        ...item,
+        uuid: typeId,
+        subItems: item.subItems?.map(mapToMenuItem),
+        label: formatMessage(labelStrings[typeId]),
+        text: textStrings[typeId] ? formatMessage(textStrings[typeId]) : undefined, // if there's a text string, show text
+      };
+    },
+    [formatMessage],
+  );
+
+  const addComponentMenuItems = useMemo(
+    () =>
+      [
+        {
+          icon: { name: 'add-plus' as IconProps.Name },
+          uuid: ObjectTypes.Component,
+        },
+        ...(isTagComponent
+          ? [
+              {
+                uuid: ObjectTypes.Overlay,
+                isDisabled: isOverlayComponent,
+                feature: { name: COMPOSER_FEATURES.Overlay },
+              },
+            ]
+          : []),
+        ...(dataBindingComponentEnabled
+          ? [
+              {
+                uuid: ObjectTypes.DataBinding,
+              },
+            ]
+          : []),
+      ].map(mapToMenuItem),
+    [selectedSceneNodeRef, isOverlayComponent, isTagComponent, dataBindingComponentEnabled],
+  );
```

Comment:
Non-blocker: This one is a bit hard to follow. Please consider refactoring to improve readability.

Reply:
ok